### PR TITLE
Added support for should queries

### PR DESCRIPTION
**Pull request by lorenzo**
This marks the last feature I wanted for this branch of the repo. Once this is merged in, I will moved the code to the master branch and start working on a branch that will have elastic 5.x as a requirement, as it is good difficult to support the 2 major versions at the same time.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -18,13 +18,13 @@
...
     "require": {
-        "ruflin/elastica": "~3.1",
-        "cakephp/cakephp": "~3.2"
+        "ruflin/elastica": "~3.1"
     },
     "require-dev": {
         "cakephp/cakephp-codesniffer": "dev-master",
         "psr/log": "~1.0",
-        "phpunit/phpunit": "~4.0"
+        "phpunit/phpunit": "~4.0",
```
Comment (markstory) [inline composer.json]: Should this go up to `~5.7`?

File: composer.json [hunk 2]
```diff
@@ -18,13 +18,13 @@
...
+        "cakephp/cakephp": "3.3.*"
```
Comment (markstory) [inline composer.json]: What about `>=3.3.0 <4.0.0`?

File: config/bootstrap.php [hunk 3]
```diff
@@ -13,26 +13,14 @@
...
-        if ($controller) {
-            $callback = ['Cake\ElasticSearch\TypeRegistry', 'get'];
-            $controller->modelFactory('ElasticSearch', $callback);
-            $controller->modelFactory('Elastic', $callback);
-        }
-    }
-);
+$callback = ['Cake\ElasticSearch\TypeRegistry', 'get'];
+FactoryLocator::add('Elastic', $callback);
+FactoryLocator::add('ElasticSearch', $callback);
```
Comment (markstory) [inline config/bootstrap.php]: :clap:

File: src/Query.php [hunk 4]
```diff
@@ -577,4 +599,42 @@ public function compileQuery()
...
+    {
+        $connection = $this->_repository->connection();
+        $name = $this->_repository->name();
+        $type = $connection->getIndex()->getType($name);
+
+        $query = clone $this->compileQuery();
+        $query->setSize(0);
+        $query->setSource(false);
+
+        return $type->search($query)->getTotalHits();
```
Comment (markstory) [inline src/Query.php]: Nice way to do get a count :+1:

File: composer.json (hunk 2 above)
Comment (lorenzo) [inline composer.json]: Currently I can't until 3.4.2 is out, since the current version has a problem with phpunit

File: composer.json (hunk 1 above)
Comment (lorenzo) [inline composer.json]: Yes

File: composer.json (hunk 2 above)
Comment (markstory) [inline composer.json]: Ok. I'm planning on doing the 3.4.2 release this week.

**Comment by lorenzo** [comment]
This can finally be merged!

**Pull request merged**